### PR TITLE
✨ New previewing engine

### PR DESCRIPTION
**Pull request by markmehere**
Please review and merge #439 and #440 first as this will reduce the number of file changes for this PR.

This PR overhauls ct.js's previewing engine. This seems a strange new feature but there are a few reasons for it:

* The new room previewer fixes a few bugs with room previewing (see #439)
* It removes 250+ lines from the core editor files 🎉
* It creates a new `prev` folder that stores all previews allowing easier git commits as this can be excluded from the repo
* It automatically regenerates the previews upon project load if not present (necessary for new git downloads)
* The style previewing has been improved so as to be between 12px and 96px as opposed to being very small and grainy or bigger than the preview window
* It deprecates the small non-retina previews as most computers will handle 128 to 320px images without missing a beat these days
* It introduces a new `trash` folder that takes images and previews that are no longer used. Files are only moved here *if there is no recovery file*, to avoid trashing files that are still in used by the "other" project file.
* It adds a background to style previews if the style is white or a near-white colour (with no alpha channel)

Skeletons are supported but unused sounds are not trashed yet. Leading to at most three pieces of further work:

1. Update the examples to the new previewing model
2. Incorporate unused sounds into the smart-trash functionality

I'm hoping this is mostly compatible with the unified asset viewer.

### CODE REVIEW

**Comment by CosmoMyzrailGorynych** [comment]
That's a lot of work and a great under-the-hood improvement!
I think we should move to working in v4 branch soon, or more git diversions will appear. I will assist with it whenever you need me (we can do a voice chat in Discord or elsewhere)

**Comment by markmehere** [comment]
v4 sounds good!

I'm pretty difficult to catch for voice calls but I agree talking a bit will probably be smoother than just springing a big pull request on you. I'll try to drop into the discord a bit more too.

I actually have a fair few pull requests that don't make it to maturity - so I never know the ones that I'll finish to think through and explain. I think maybe having a Github wiki or something similar could help for co-ordination. Here's a [wiki page](https://github.com/markmehere/ct-js/wiki/ct%E2%80%90js-wishlist) on my fork that sort of outlines my wishlist for ct-js.

This PR should be mature enough to merge. You may get some obsolete thumbnails when switching between the recovery and project file but otherwise should be good (you could always regenerate previews to avoid this but that seems quite expensive).

My contributions to ct-js will probably be pretty sporadic as I kind of like to use it for programming games too. 😁

**Comment by markmehere** [comment]
~Actually wait a minute I might add a background for rooms just noticed this!~ 😁 Sorted

**Comment by CosmoMyzrailGorynych** [comment]
Will start merging it into pixi-v7 now.
I also see that you added background color here, too. (I would go with PIXI.Graphics too lol)

**Comment by CosmoMyzrailGorynych** [comment]
Also please install ESLint extension to your code editor, by gods there is a ton of auto-fixable problems and places to compute in parallel.

File: src/node_requires/resources/preview/skeleton.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { imageContain, toBuffer } from "../../utils/imageUtils";
+import { getDOMImage } from "../textures";
+
+/*
+ * Dragon bones is like my hockey team whenever I play - no support whatsoever.
+ */
```
Comment (CosmoMyzrailGorynych) [inline src/node_requires/resources/preview/skeleton.ts:6]: ct.js v4 will use Spine as its renderer, DragonBones is just dead as a framework. But is perfectly usable as an editor, I added an automatic converter from DB to Spine to ct.js.

File: src/node_requires/resources/preview/index.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    fs.ensureDir(global.projdir + "/prev");
+    for (const font of projectData.fonts) {
+        if (!fs.existsSync(FontPreviewer.get(font, true))) {
+            await FontPreviewer.save(font);
+            previewGenerated = true;
+        }
+    }
+    for (const room of projectData.rooms) {
+        if (!fs.existsSync(RoomPreviewer.get(room, true))) {
+            await RoomPreviewer.save(room, room === projectData.rooms[0]);
```
Comment (CosmoMyzrailGorynych) [inline src/node_requires/resources/preview/index.ts:72]: It should use `project.startroom`, then default to the first room.

Comment (CosmoMyzrailGorynych) [reply src/node_requires/resources/preview/index.ts:72]: I created a new method `getStartingRoom` in `node_requires/resources/rooms` as it is also needed in the room editor.

**Comment by CosmoMyzrailGorynych** [comment]
"Merged" in https://github.com/ct-js/ct-js/commit/14e13b0062cc7a43f3e3e0392b6a9a49a3777151 🐙 

Changes:
* Updated for ct.js v4 folder structure with pixi.js v7 changes addressed
* Previews are generated in parallel asynchronously
* Somehow I skipped `resources/styles` or they are missing in PR completely. I fixed it. Your previewer is in the commit.
* Changed splash screen generation logic so it uses the real starting room.
* UI rooms won't have a solid background in the preview.
* Added .gitignore to new projects.
* Added .gitignore rules for ct-js repo.

**Comment by CosmoMyzrailGorynych** [comment]
изображение

**Comment by CosmoMyzrailGorynych** [comment]
@markmehere The catformer example has a preview issue with tiled backgrounds:

изображение

(Maybe that's I who goofed up, I had been incorporating all these changes for good 5 hours and could have skipped something)

**Comment by CosmoMyzrailGorynych** [comment]
On the side note, the `pixi-v7` branch is close to being merged into `develop`, so if you have other improvements and features in mind, I recommend switching to it for development. It lacks sound editor completely atm, but otherwise is a new shiny working engine. And sounds overhaul will be merged soon anyways.

**Comment by markmehere** [comment]
Lots of comments. I shall take a look. ~I thought tiles were working well.~ 🤔

Maybe not. I'll see if I can shift to pixi-v7. I agree with getting rid of DragonBones. I couldn't get it working in the end.

**Pull request closed without merging**